redbot/cogs/alias/alias_entry.py

```diff
@@ -30,34 +42,182 @@ def inc(self):
         ret.uses = data.get("uses", 0)
         return ret
+
+
+class AliasCache:
+    def __init__(self, config: Config, cache_enabled: bool = True):
+        self.config = config
+        self._cache_enabled = cache_enabled
+        self._loaded = False
+        self._aliases: Dict[Optional[int], Dict[str, AliasEntry]] = {None: {}}
+
+    async def load_aliases(self):
+        if not self._cache_enabled:
+            self._loaded = True
+            return
+        for alias in await self.config.entries():
+            self._aliases[None][alias["name"]] = AliasEntry.from_json(alias)
+
+        all_guilds = await self.config.all_guilds()
+        async for guild_id, guild_data in AsyncIter(all_guilds.items(), steps=100):
+            if guild_id not in self._aliases:
+                self._aliases[guild_id] = {}
+            for alias in guild_data["entries"]:
+                self._aliases[guild_id][alias["name"]] = AliasEntry.from_json(alias)
+        self._loaded = True
+
+    async def get_aliases(self, ctx: commands.Context) -> List[AliasEntry]:
+        """Returns all possible aliases with the given context"""
+        global_aliases: List[AliasEntry] = []
+        server_aliases: List[AliasEntry] = []
+        global_aliases = await self.get_global_aliases()
+        if ctx.guild and ctx.guild.id in self._aliases:
+            server_aliases = await self.get_guild_aliases(ctx.guild)
+        return global_aliases + server_aliases
+
+    async def get_guild_aliases(self, guild: discord.Guild) -> List[AliasEntry]:
+        """Returns all guild specific aliases"""
+        aliases: List[AliasEntry] = []
+
+        if self._cache_enabled:
+            if guild.id in self._aliases:
+                for _, alias in self._aliases[guild.id].items():
+                    aliases.append(alias)
+        else:
+            aliases = [AliasEntry.from_json(d) for d in await self.config.guild(guild).entries()]
+        return aliases
+
+    async def get_global_aliases(self) -> List[AliasEntry]:
+        """Returns all global specific aliases"""
+        aliases: List[AliasEntry] = []
+        if self._cache_enabled:
+            for _, alias in self._aliases[None].items():
+                aliases.append(alias)
+        else:
+            aliases = [AliasEntry.from_json(d) for d in await self.config.entries()]
+        return aliases
+
+    async def get_alias(
+        self, guild: Optional[discord.Guild], alias_name: str,
+    ) -> Optional[AliasEntry]:
+        """Returns an AliasEntry object if the provided alias_name is a registered alias"""
+        server_aliases: List[AliasEntry] = []
+
+        if self._cache_enabled:
+            if guild is not None:
+                if guild.id in self._aliases:
+                    if alias_name in self._aliases[guild.id]:
+                        return self._aliases[guild.id][alias_name]
+            if alias_name in self._aliases[None]:
+                return self._aliases[None][alias_name]
```

Comment:
These need to be opposite order - global aliases take precedence over guild aliases.
```suggestion
            if alias_name in self._aliases[None]:
                return self._aliases[None][alias_name]
            if guild is not None:
                if guild.id in self._aliases:
                    if alias_name in self._aliases[guild.id]:
                        return self._aliases[guild.id][alias_name]
```

​

If you care for my nitpicks, this code would avoid doing key lookups multiple times, otherwise just ignore this part, I couldn't stop myself from putting this here :P
```suggestion
            if alias := self._aliases[None].get(alias_name):
                return alias
            if guild is not None:
                if guild_aliases := self._aliases.get(guild.id):
                    if alias := guild_aliases.get(guild.id):
                        return alias
```